src/ric/d_18C40.c

```diff
@@ -0,0 +1,330 @@
+    0xFFFD0000,
+    0xFFFDAAAB,
+};
+
+s32 D_80154ED4[] = {
+    0x00010004, 0x00020004, 0x00030004, 0x00040004, 0x00050004,
+    0x00060004, 0x00070004, 0x00080004, 0x00000000,
+};
+
+s32 D_80154EF8[] = {
+    0x00090008, 0x000D0008, 0x000A0040, 0x000B0002, 0x000C0002, 0x000D0006,
+    0x000E0007, 0x000F0006, 0x000E0005, 0x000D0004, 0x000F0003, 0x000E0003,
+    0x000D0003, 0x000E0003, 0x000F0003, 0x000E0003, 0x000D0003, 0x000E0004,
+    0x000F0005, 0x000E0006, 0x000D0007, 0x000E0030, 0x0009000C, 0x0010000D,
+    0x00110008, 0x0012000C, 0x001300B0, 0x0014000A, 0x0015000A, 0x0016000A,
+    0x00170030, 0x001800D0, 0x0000FFFF,
+};
+
+s32 D_80154F7C[] = {
+    0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000,
+    0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000,
+    0x00000000, 0x00000000, 0x00000000, 0x00000000,
+};
+
+s16 D_80154FBC[][10] = {
+    0x0000, 0x0000, 0x0000, 0x015F, 0x0100, 0x0100, 0x0100, 0x0060, 0x0031,
+    0x0200, 0x0000, 0x0002, 0x0006, 0x015F, 0x0100, 0x0100, 0x0100, 0x0040,
+    0x0031, 0x0200, 0x0000, 0x0000, 0x0000, 0x015F, 0x0100, 0x7000, 0x7000,
+    0x0040, 0x0031, 0x0200, 0x0000, 0x0000, 0x0000, 0x015F, 0x0100, 0x7000,
+    0x0200, 0x0010, 0x0031, 0x0200, 0x0000, 0x0000, 0x0000, 0x015F, 0x0100,
+    0x7000, 0x7000, 0x0010, 0x0031, 0x0200, 0x0000, 0x0000, 0x0000, 0x015F,
+    0x0100, 0x7000, 0x0200, 0x7000, 0x0031, 0x0200, 0x0000, 0x0000, 0x0000,
+    0x015F, 0x0100, 0x7000, 0x0100, 0x7002, 0x0031, 0x0100, 0x0000, 0x0000,
+    0x0000, 0x015F, 0x0100, 0x0100, 0x7000, 0x0070, 0x0031, 0x0200, 0x0000,
+    0x0000, 0x0000, 0x015F, 0x0100, 0x7000, 0x0200, 0x7006, 0x0031, 0x0200,
+    0x0000, 0x0000, 0x0000, 0x015F, 0x0200, 0x0200, 0x0400, 0x7001, 0x0031,
+    0x0200, 0x0000, 0x0000, 0x0000, 0x0166, 0x0100, 0x0100, 0x0100, 0x7005,
+    0x0031, 0x0200, 0x0000, 0x0000, 0x0000, 0x0102, 0x0100, 0x7000, 0x7000,
+    0x0040, 0x0031, 0x0200, 0x0000, 0x0000, 0x0000, 0x015F, 0x0100, 0x7100,
+    0x0100, 0x7005, 0x0031, 0x0200, 0x0000, 0x0000, 0x0000, 0x015F, 0x0100,
+    0x0100, 0x0100, 0x7005, 0x0031, 0x0200, 0x0000, 0x0000, 0x0000, 0x0102,
+    0x0100, 0x7000, 0x0200, 0x0040, 0x0031, 0x0200, 0x0000, 0x0000, 0x0000,
+    0x0102, 0x0100, 0x7000, 0x0100, 0x0040, 0x0031, 0x0200, 0x0000, 0x0000,
+    0x0000, 0x0102, 0x0100, 0x0100, 0x7000, 0x0000, 0x0031, 0x0200, 0x0000,
+    0x0000, 0x0000, 0x0102, 0x0400, 0x0100, 0x0400, 0x0000, 0x0031, 0x0200,
+    0x0000, 0x0000, 0x0000, 0x015F, 0x0100, 0x0100, 0x7000, 0x0000, 0x0031,
+    0x0200, 0x0000, 0x0000, 0x0000, 0x015F, 0x0200, 0x0800, 0x0800, 0x7007,
+    0x0031, 0x0200, 0x0000, 0x0000, 0x0000, 0x0102, 0x0100, 0x0100, 0x0100,
+    0x0000, 0x0031, 0x0200, 0x0000, 0x0000, 0x0000, 0x015F, 0x00C0, 0x00C0,
+    0x0100, 0x0010, 0x0031, 0x0200, 0x0008, 0x0000, 0x0008, 0x0165, 0x0100,
+    0x0100, 0x0100, 0x7007, 0x0051, 0x0200, 0x0000, 0x0000, 0x0000, 0x0168,
+    0x7100, 0x7100, 0x0200, 0x7007, 0x0031, 0x0200,
+};
+
+Point16 D_8015519C[] = {
```

Comment:
This and  `D_80155244` seem out of sync despite referring to the same data? I'm thinking D_8015519C may be shorter than this and only goes up to before 0x18
```
    loadedParams = &D_8015519C[upperparams >> 8];
    rScale = loadedParams[2];
    gScale = loadedParams[3];
    bScale = loadedParams[4];
    gOffset = loadedParams[0];
    bOffset = loadedParams[1];
```
It would be good to check what the possible values for upperparams are.

Reply:
> It would be good to check what the possible values for upperparams are.

I did some analysis, here's my full thinking. Skip to the bottom for my personal conclusion on the size.

The EntityShrinkingPowerUpRing is made by an entity factory.

EntityEntFactory controls its child's params like this:

newEntity->params = self->ext.factory.unkA0

if self->ext.factory.unkA2 != 0{
    newEntity->params += self->ext.factory.unkA6
} else {
    newEntity->params += i
}

First let's look at that second portion. What's unkA2? That comes from the blueprint. We're using RIC blueprints 36 through 39.
unkA2 comes from the start of EntityEntFactory, and appears to be unk2, bit 6. For those blueprints, that value is TRUE. So we're doing += unkA6. That doesn't appear to be set anywhere, so it would always be zero. So all we worry about is the unkA0.
That directly comes from the FACTORY() call. If we use FACTORY(0x400), then the upperparams becomes 0x400.

For calls to FACTORY in RIC, I've searched for blueprints 36 through 39, and it appears that they are only 0, 0x100, 0x200, 0x300, and 0x400.

This indicates that D_8015519C has a length of 5 or more, since we're using elements 0, 1, 2, 3, and 4.

It's possible I missed something somewhere in this analysis, but that's at least my starting point for anyone else to build on.

Reply:
Yes, this is definitely wrong, it was a pain to make it work tho, you have to be very careful with the array and struct sizes otherwise the compiler will generate padding for the missing data. I'll take a second look.